lib/assignment_pipeline.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+# Manages the current status of an assignment.
+class AssignmentPipeline
+  module AssignmentStatuses
+    IN_PROGRESS = 'in_progress'
+    READY_FOR_REVIEW = 'ready_for_review'
+    READY_FOR_MAINSPACE = 'ready_for_mainspace'
+  end
+
+  ASSIGNMENT_STATUS_PIPELINE = [
+    AssignmentStatuses::IN_PROGRESS,
+    AssignmentStatuses::READY_FOR_REVIEW,
+    AssignmentStatuses::READY_FOR_MAINSPACE
+  ]
+
+  module ReviewStatuses
+    PEER_REVIEWER_NEEDED = 'peer_reviewer_needed'
+    PEER_REVIEW_STARTED = 'peer_review_started'
+    PEER_REVIEW_COMPLETED = 'peer_review_completed'
+  end
+
+  REVIEW_STATUS_PIPELINE = [
+    ReviewStatuses::PEER_REVIEWER_NEEDED,
+    ReviewStatuses::PEER_REVIEW_STARTED,
+    ReviewStatuses::PEER_REVIEW_COMPLETED
+  ]
+
+  def initialize(flags:)
+    @flags = flags
+  end
+
+  def assignment_status
+    last_status(@flags[:assignment_status]) || ASSIGNMENT_STATUS_PIPELINE.first
+  end
+
+  def next_assignment_status
+    next_status(ASSIGNMENT_STATUS_PIPELINE, assignment_status)
+  end
+
+  def increment_assignment_status
+    @flags[:assignment_status] ||= []
```

Comment:
`@flags` here is going to be from an `Assignment` instance? So this is mutating the flags object that got passed in, but leaving it up to `Assignment` to then save it? Might be better to pass in the whole assignment to this class so that the flags can be both changed and persisted together, or alternatively, don't do any mutations of `@flags` and leave that to the code that calls this class.